### ISSUE TITLE
[RFC/Design] Standardize probabilistic vs deterministic return contract to unblock evaluation integrations (Can be Taken under #49 / #62 )

### ISSUE THREAD

**Issue by kshirajahere** [mllam/neural-lam#335]
## Background
As probabilistic support evolves, parts of the current model/eval flow still assume a single deterministic prediction tensor.

This makes integration brittle:
- deterministic and probabilistic branches return different shapes/interfaces,
- evaluation/plotting logic has to infer behavior ad-hoc,
- adding eval-only probabilistic diagnostics becomes harder.

## Proposal

Introduce a small, explicit forecast output contract used by `Forecaster` / `StepPredictor`:

- deterministic path remains supported without behavior change
- probabilistic models can optionally return sampled forecasts via a dedicated field

Example structure (name flexible, e.g. `ForecastOutput`):
- `point_prediction`: `(B, T, N, F)` (required)
- `samples`: optional `(B, S, T, N, F)`
- `pred_std`: optional `(B, T, N, F)`
- `aux`: optional dict for model-specific extras

## Non-breaking rollout

1. Add the output type + shape validator utility.
2. Update `ForecasterModule` to accept either:
   - legacy tensor output (current behavior), or
   - new structured output.
3. Keep training/eval behavior unchanged for deterministic models.
4. Add tests for:
   - legacy deterministic compatibility
   - probabilistic sample path shape validation
   - clear error messages on invalid dimensions

## Related

Refs #49  
Refs #62

## Architecture Diagram of Solution

<p align="center">
  <img width="700" height="2000" src="https://github.com/user-attachments/assets/ce31037d-27e4-4135-b2f3-d2f2d4fdaf40" />
</p>

## Notes
Out of scope: implementing new probabilistic model architectures in this issue.

Would love maintainer guidance on which structure fits current `main` direction best:
- `Dict`
- `dataclass`
- native tensors + metadata

**Comment by abhaygoudannavar** [comment]
hey @kshirajahere i was looking at the issue nice RFC , but Quick question on the samples: (B, S, T, N, F) field — the current AR loop propagates a single state forward at each step. If you have S ensemble samples, each sample diverges at every AR step, so you'd need S parallel rollouts, not just a reshape at the output. How do you envision the unrolling loop handling this? It seems like more of an architectural change to ARForecaster than an output format change.

**Comment by kshirajahere** [comment]
> If you have S ensemble samples, each sample diverges at every AR step, so you'd need S parallel rollouts, not just a reshape at the output

Hmmmm, you’re correct that true stochastic AR sampling needs `S` independent state trajectories during rollout (not a post-hoc reshape).

That’s exactly why **this** issue is scoped as an interface RFC, not an ARForecaster implementation PR.

The separation is intentional:

- This issue: define a stable output contract (`point_prediction`, optional `samples`, etc.) and validation so metrics/plots stop depending on ad-hoc shape assumptions.
- Follow-up architecture work (#49/#62): implement sample-producing paths (sample-aware AR unroll, repeated seeded unroll + stack, or model-native sample emission).

Without a contract first, producer-side work keeps forcing downstream API churn.  
With a contract first, producer implementations can evolve independently while consumers remain stable.

If the organisation maintainers, prefer a explicit distinction.. I can edit out the issue anytime :) Thank you for the inputs @abhaygoudannavar

**Comment by Saptami191** [comment]
@abhaygoudannavar  That's a good point, and yes, it would require S independent rollout trajectories for (B,S,T,N,F) in an autoregressive model.I suppose the purpose of this RFC is mostly to standardize the output contract so that it doesn't depend on any assumptions about the shape of the tensor. In a deterministic pipeline, `samples` would simply remain None.
Any implementation that actually generates samples would require changes to the AR rollout, which would probably be part of the discussions around the forecaster/model refactor in #49 / #62.

I am already working on support area  agnostic graph generation issue . @kshirajahere  r u working on it ?

**Comment by kshirajahere** [comment]
@Saptami191 exactly the point of this RFC is to standardise the output contract.
on your question: I’m currently focused on issue #335 (output contract RFC) and my existing WeatherDataset PR threads (#312 / #200 context), not on area-agnostic graph generation(haven't gone through yet).

But I will surely get back to you on #299 with my thoughts and reviews soon(if you want ofc).

**Comment by Saptami191** [comment]
@kshirajahere thank u . I just suggested that the issue  #299 helps a little. I have already solved that issue .just need a review

**Comment by kshirajahere** [comment]
@Saptami191 cool i will see thru it thanks :)

**Comment by techaadii** [comment]
Well I would love to help on this issue

**Comment by joeloskarsson** [comment]
Hi, thanks for starting a discussion on this! I agree with some comments above that we need to think about how the unrolling happens here, and that we never unroll over multiple samples (ensemble members) at once. While the diagram is useful, I found it a bit hard to understand what each arrow means, so apologies if I misunderstood something.

One thing that we can surely decide is the tensor shapes:
* A deterministic (single) forecast: (B, T, N, F), optionally with matching std-dev of (B, T, N, F)
* An ensemble forecast: (B, S, T, N, F), optionally with matching std-dev of (B, S, T, N, F) (if that is something we want, then becomes a mixture of Gaussians)

This should be enough in itself to implement any plotting and evaluation routines right? There are few metrics where it makes sense to feed either a deterministic or ensemble forecast, and often it is the responsibility of whoever calls e.g. RMSE to decide that they want it for the ensemble mean. If a metric should be available for both cases (e.g. log-spectral-distance), then it can check its input if it is a single forecast or ensemble. If you don't find tensor shape robust enough for this we could wrap these forecasts some way, but this wrapper should be _either_ a deterministic forecast or ensemble. My understanding from above (pls explain if I misunderstand) is that this construct could include both in one `ForecastOutput`.

For re-using StepPredictor and ARForecaster (based on #208) also for ensemble models, in my view it would make most sense to have StepPredictor always be a mapping from one or more (batch of) past atmospheric states to another, e.g. 2x(B, N, F) --> (B, N, F). Then this can be either a deterministic or stochastic mapping. Then the ARForecaster will always return a batch of forecasts of shape (B, T, N, F), and it is up to whoever uses it to parallelize ensemble forecasting. For paralellizing this on one GPU, one can always fold the ensemble dim into the batch dim, or we might even have ARForecaster accept arbitrary batch dimensions.

**Comment by kshirajahere** [comment]
I think I was overloading the RFC by allowing one ForecastOutput to carry both point_prediction and samples.

Your formulation makes sense to me:
- deterministic forecast: (B, T, N, F), optionally with matching pred_std
- ensemble forecast: (B, S, T, N, F), optionally with matching pred_std if we explicitly want per-member Gaussian components

I agree that this is likely enough for plotting/evaluation on its own, with the caller deciding when to reduce an ensemble to a point forecast, and only metrics that genuinely support both cases checking whether the input is deterministic or ensemble.

I also agree with the #208  direction that StepPredictor and ARForecaster should stay single-trajectory in their core semantics:
- StepPredictor: past state(s) + forcing ---> next state, optionally with pred_std
- ARForecaster: unroll one forecast of shape (B, T, N, F)

Then ensemble generation can sit outside that path by parallelizing over samples, either by folding S into the batch dimension or by letting the forecaster accept arbitrary leading batch dimensions. That feels cleaner than making ARForecaster itself own a special ensemble axis.

So I think the right revision from my interpretation of things would be to drop the combined ForecastOutput(point_prediction, samples, ...) idea and instead define a minimal forecast contract around standardized tensor shapes, with an optional thin wrapper only if we need an explicit deterministic-vs-ensemble tag.

**Comment by Sir-Sloth-The-Lazy** [comment]
Thanks @joeloskarsson , this clarifies the design significantly. The direction of keeping ARForecaster as a single-trajectory forecaster returning (B, T, N, F) and folding S into B at the caller level makes sense to me, and I'll update my GSoC proposal architecture accordingly.
One question on the stochastic case: in Graph-EFM, each ensemble member requires a freshly sampled Z_t from LatentMap at every step. If S is folded into B before calling ARForecaster, then StepPredictor.forward() receives a batch of size B*S  but LatentMap needs to know which entries in that batch are different members (to sample independent Z_t per member) vs. different batch items. Is the intent that StepPredictor handles this internally (e.g. HierarchicalGraphEFM) samples a fresh Z_t independently for every item in whatever batch it receives making the fold transparent? If so I think that works cleanly, just want to confirm before I commit to that in the proposal.

**Comment by Sir-Sloth-The-Lazy** [comment]
another question @joeloskarsson ,following your comment here about keeping ARForecaster as a single-trajectory forecaster and handling ensemble generation via batch-dim folding, I want to make sure I understand the intended design correctly before finalising my proposal.
My current thinking is to keep ARForecaster unchanged and have ForecasterModule handle the fold/unfold logic for ensemble generation. But I'm wondering: would it make sense to have a thin EnsembleForecaster class that owns just the fold/unfold logic (not a separate rollout loop), so callers don't repeat that pattern? Or is the preference to keep it explicitly in calling code and not introduce a new class at all?
Asking because HierarchicalGraphEFM and a future DiffusionStepPredictor would both benefit from the same ensemble wrapper if one existed.

**Comment by PranjaliAwchar** [comment]
I’ve been going through the RFC and the points raised here, and the proposed tensor shape contract makes a lot of sense. Having deterministic forecasts as (B, T, N, F) and ensemble forecasts as (B, S, T, N, F) feels like a clean way to separate concerns, especially since StepPredictor and ARForecaster remain single‑trajectory while ensemble generation is handled externally. That matches well with the direction in #208 and also connects closely to my GSoC proposal idea (#62) on probabilistic forecasting.
From a contributor’s perspective, I’d be interested in helping with the rollout steps mentioned — for example, adding shape validation utilities, updating plotting/evaluation routines to consume these standardized shapes explicitly, and writing tests to cover both deterministic and ensemble cases. I’m also curious about how evaluation metrics like CRPS could branch depending on whether the input is deterministic or ensemble.

**Comment by kshirajahere** [comment]
Great points and questions, @Sir-Sloth-The-Lazy! Regarding your first question on the $Z_t$ sampling: flattening $S$ into the batch dimension (B * S) actually handles the stochastic sampling perfectly natively. Because pytorch draws random tensors based purely on the first dimension size, LatentMap drawing noise for `shape[0]` (which is now B*S) guarantees every single ensemble member across every batch item receives a uniquely sampled $Z_t$. The StepPredictor essentially remains entirely agnostic to whether it's processing different geographical batches or stochastic permutations of the same batch.

Regarding your second point, I am completely in agreement. In my GSoC proposal draft for Project 3, I actually explicitly mapped this out using a lightweight wrapper pattern (something functionally akin to your suggested EnsembleForecaster). Rather than bloating ForecasterModule with complex `reshape()` and unfolding logic, a thin decorator or composition wrapper class around ARForecaster allows Graph-EFM and future Diffusion branches to easily trigger multi-trajectory rollouts without ARModel or the main validation steps needing to change their (B, T, N, F) expectations. @joeloskarsson, assuming the math on the independent $Z_t$ distributions above aligns with the current Graph-EFM bounds, do you prefer a dedicated wrapper class for the unroll mapping as suggested, or would you rather contain the shape manipulation strictly inside pytorch lightning's validation_step?

**Comment by Sir-Sloth-The-Lazy** [comment]
Thanks @kshirajahere for the perspective on the Z_t sampling, that matches my intuition as well. Will wait for @joeloskarsson's take on the wrapper class question before finalising the proposal architecture.

**Comment by joeloskarsson** [comment]
@Sir-Sloth-The-Lazy  Regarding folding in the batch-dim and how to deal with the LatentMap: The simplest solution is to just always treat incoming samples in the batch independently, no matter if they come from shared initial conditions or not. Z_t should always be sampled independently for all of these. However, we could technically only compute the prior distribution by running the forward pass of LatentMap once per "true" sample in the batch (shared among ens members). But I am not sure if the computational gain is worth it for the implementation complexity. If we would allow for arbitrary leading batch dimensions then this could be done easier, but still not obvious how to do the best way (if we get an initial condition of shape (B,S,2,N,F) to the LatentMap, how do we know that the S different members have the same initial conditions, and we can safely only compute the prior once and reuse? This will be true for time step 0 but not later).

> would it make sense to have a thin EnsembleForecaster class that owns just the fold/unfold logic 
yes, we should avoid duplicating this. I would think that this could be handled by extending the ForecasterModule to an EnsForecasterModule handling this and other ensemble-specific things. But maybe it could be done by extending the Forecaster as well.

**Comment by kshirajahere** [comment]
yes this helps a lot. my read now is that #335 is converging on:

- `StepPredictor` stays single-trajectory
- `ARForecaster` stays single-trajectory
- ensemble generation sits outside that core path
- `LatentMap` just treats every incoming batch item independently and samples `Z_t` independently

which seems like the cleanest way to keep the deterministic hierarchy reusable for project 3. for the fold/unfold part, i also agree it would be better to centralize that somewhere instead of letting each probabilistic model do it ad hoc. `EnsForecasterModule` sounds like a pretty reasonable place for that to live.

**Comment by kshirajahere** [comment]
@joeloskarsson on the computational question, my bias would be to keep the first implementation fully batch-agnostic and treat every folded item independently, without trying to infer which batch indices “belong together” for shared-prior reuse.

That seems like the cleanest way to preserve the current direction from this RFC:
- `StepPredictor` stays single-trajectory
- `ARForecaster` stays single-trajectory
- ensemble generation is handled outside that core path

If repeated prior computation later turns out to be a real bottleneck, I think that optimization is better introduced explicitly (for example via a grouping-aware interface or predictor-specific override) rather than inferred from tensor shape alone.

**Comment by Sir-Sloth-The-Lazy** [comment]
@joeloskarsson  One thing that struck me while reviewing this conversation is that the redundant computation of the prior from the batch-dim folding is actually limited to t = 0 anyway, because we're actually using a GraphLatentEncoder / HiGraphLatentEncoder, which depends on the input state, and all members of S share the same initial state so we're computing S identical priors, but from t = 1 onwards, they're all different because they've all received a different Z_t. And for a rollout of any significant length, this is one redundant computation out of many. When `learn_prior = False`, we're actually using a `ConstantLatentEncoder` that just `returns torch.ones` regardless of input, so this is actually the case for all time steps, but this computation is obviously very fast anyway. So I think we're actually right to start from the **fully batch-agnostic approach** as @kshirajahere suggested, fold S into B, let LatentMap treat all members as independent, and let all three components, LatentMap, StepPredictor, ARForecaster, be completely ensemble-unaware. The folding/unfolding should all stay inside `EnsForecasterModule`. If we find that this computation is actually a performance bottleneck, e.g., HiGraphLatentEncoder with many levels, we can actually fix this without modifying the interface, by splitting `BaseLatentEncoder.forward` into `compute_distribution`, which `returns the Normal`, and `sample`, which actually draws from this Normal, so we can compute the Normal once per actual item of the batch and then use `rsample((S,))` to get all members of the ensemble at time 0. This would be backward-compatible without having to modify LatentMap, which would not have to know anything about the ensemble. As for whether we actually need a way to tell StepPredictor which batch indices belong together... I think we actually don't. The EnsForecasterModule is the one that actually folded S into B, so we know this information anyway. We'd be adding interface complexity for a performance improvement that's actually limited to a single time step.

**Comment by kshirajahere** [comment]
@joeloskarsson that matches my read as well. I agree the first slice should stay fully batch-agnostic and keep the ensemble-unaware path clean, with any shared-prior optimization treated as a later performance improvement rather than something the interface needs to encode up front.

**Comment by sadamov** [comment]
Closing as the probabilistic vs deterministic return contract design is now folded into the consolidated #649 (probabilistic metrics & ensemble plumbing). The API contract discussion can continue there.
